### PR TITLE
backup: add --dry-run/-n flag

### LINKED ISSUE #1542: Dry-Run Backup
Related to #10

It would be nice to be able to see what all will be changed via a snapshot before actually creating one. In essence, a way to show a diff between the files stored in restic, and a given path on the local filesystem. This would especially help when making changes to backup configurations so as to be able to catch anything that might be missing/errantly included before hand.

I assume given the architecture, this would be impossible to do without going through the steps of an actual backup. If so, I think it makes more sense to position it as a dry-run backup (where nothing is actually committed) rather than a diff a la `git status`.

### PR DESCRIPTION
**Pull request by aawsome**
What does this PR change? What problem does it solve?
-----------------------------------------------------

This PR is a follow-up of #2308 and #3210 as both PRs seem to be stale.
I used the changes by @rmmh and @erincerys and did the following changes:
- simplified the dry run backend quite a lot (but is now only working for backup dry runs)
- changed the tests accordingly and separated the integration test for the dry run `backup` run

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

closes #1542

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by JoshStrobl** [comment]
Just added this patch to the restic package for Solus, was rather surprised this functionality wasn't already in it. Works as intended.

File: doc/040_backup.rst
Comment (obbardc) [inline doc/040_backup.rst:198]: why does `--dry-run` have to be combined with `--verbose`?

File: doc/040_backup.rst
Comment (obbardc) [inline doc/040_backup.rst:202]: you say that `--dry-run` has to be combined with `--verbose`, yet you combine it with `-vv`

File: doc/040_backup.rst
Comment (aawsome) [reply doc/040_backup.rst:198]: It does not have to be combined. `backup --dry-run` gives the same output as `backup` without dry-run (except it replaces the sentence "Added to the repo" by "Would be added to the repo"). But if you want to see the changes by details, you have to run it with increased verbosity.

If you think that this part of the docu is unclear: Do have a suggestion how to describe it?

File: doc/040_backup.rst
Comment (aawsome) [reply doc/040_backup.rst:202]: `-vv` equals to `--verbose=2`

File: doc/040_backup.rst
Comment (abathur) [reply doc/040_backup.rst:198]: This may just be a little linguistic confusion around "do not **write** anything"?

A few alt takes:
- `report what would be done without modifying the repository`
- `report what would be done, but do not modify the repository`
- `report what would be done without doing it`
- `report what would be done without doing anything`
- `report what would be done, but do nothing`
- `report what would be done, but make no changes`

**Comment by Iodok** [comment]
Just tried the --dry-run with backup with my restic  version
restic 0.12.0 compiled with go1.15.8 on linux/amd64
My cmdline was: restic backup --dry-run -r $RESTIC_REPO $RESTIC_SRC -p $RESTIC_PASSWORDFILE 
and get:
unknown flag: --dry-run
Do you have any hints?

Iodok

**Comment by JoshStrobl** [comment]
@Iodok The patch hasn't been merged in and is awaiting changes. The process for building Restic is fairly trivial, you can apply the patch separately or wait until it has been merged in and a future release occurs with the change in it.

**Comment by bjoe2k4** [comment]
Works as intended. Please merge for inclusion in next release :-)

File: internal/repository/repository.go
Comment (MichaelEischer) [inline internal/repository/repository.go:13]: **question (non-blocking):** I wonder whether there's an elegant way to avoid the explicit dependency on the dry-run backend here. Using the `backendTestHook` would probably work, but that definitely doesn't count as elegant.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
The code seems to be in quite good shape already. The main issue I see right now is that it's currently too easy to not shadow newly introduced modifying backend operations or no longer overwrite them after signature changes. See the comment for a suggestion.

File: doc/040_backup.rst
Comment (rawtaz) [reply doc/040_backup.rst:198]: `report what would be done, without modifying the repository` :-)

File: internal/repository/repository.go
Comment (aawsome) [reply internal/repository/repository.go:13]: IMO the `backendTestHook` would not work as intended. It is applied before the the backend is wrapped by a `CachedBackend`. So, dry-run writes to the repository would not write to the repo, but write to the cache which is not intended.

The problem with the cache is that it depends on the repo ID which is only available after actually reading the config file from the backend. This is why the cache wrapper is added in `repository.go`. Now as the dryrun wrapper should be applied to the already cache-enabled backend, I added this also here.
The alternative would be to allow direct access to the backend of the repository via some `GetBackend` and `SetBackend` methods and then call
```golang 
repo.SetBackend(dryrun.New(repo.GetBackend()))
```
in `cmd_backup.go`. This way we could also separate (most of) the cache logic from the backend.

File: internal/backend/dryrun/dry_backend.go
Comment (aawsome) [reply internal/backend/dryrun/dry_backend.go]: is changed.

File: internal/backend/dryrun/dry_backend_test.go
Comment (aawsome) [reply internal/backend/dryrun/dry_backend_test.go]: It isn't necessary. I removed this filtering out.

File: internal/backend/dryrun/dry_backend.go
Comment (aawsome) [reply internal/backend/dryrun/dry_backend.go]: Actually I just realized that in order to get the length, we can simply use `rd.Length()`. Thanks for the hint. 

I also checked out how this `Save` method is actually called and decided that the dry-run backend doesn't even need to read from the reader, so I removed this and now use `rd.Length()` to display the omitted size in the debug output.
Note that this in fact means that the dry run backend does now no longer exactly behave the same way with respect to `rd` like the original backend would. (but also other backends like the `CachedBackend` or the `RetryBackend` don't).
Also this means that `backup` with the `--dry-run` flag does no longer read previously written temporary files (or memory buffers) and should be a bit faster.

File: doc/040_backup.rst
Comment (aawsome) [reply doc/040_backup.rst:198]: is changed.

File: internal/repository/repository.go
Comment (MichaelEischer) [reply internal/repository/repository.go:13]: Indeed, you're right, the backend is wrapped with the `CacheBackend` later on.

Making arbitrary modifications to the used backend also doesn't feel particularly elegant, so let's just leave it like it is now.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks for completing the work on this option!

**Comment by MichaelEischer** [comment]
I've added a short paragraph describing the use cases solved by this PR.

**Comment by MichaelEischer** [comment]
I've rebased the PR before merging.

**Pull request merged**
